pkg/provisioner/cloudprovisioner.go

```diff
@@ -377,20 +380,31 @@ func (c *CloudProvisioner) PublishVolume(
 		w.Logger().V(2).Infof("Trying to attach volume %q to node %q.", volumeID, nodeName)
 		var cachingMode compute.CachingTypes
 		if cachingMode, err = azureutils.GetCachingMode(volumeContext); err != nil {
-			return nil, err
+			return attachResultCh, nil, err
 		}
 
+		lunCh := make(chan int32, 1)
+		ctx = context.WithValue(ctx, azure.LunChannelContextKey, lunCh)
 		asyncAttach := azureutils.IsAsyncAttachEnabled(c.enableAsyncAttach, volumeContext)
-		lun, err = c.cloud.AttachDisk(ctx, asyncAttach, diskName, volumeID, nodeName, cachingMode, disk)
-		if err != nil {
-			w.Logger().Errorf(err, "attach volume %q to instance %q failed", volumeID, nodeName)
-			return nil, err
-		}
-		w.Logger().V(2).Infof("attach operation successful: volume %q attached to node %q.", volumeID, nodeName)
+		attachResultCh = make(chan azureutils.CloudAttachResult, 1)
+		go func() {
+			ctx, w := workflow.New(ctx)
+			defer func() { w.Finish(err) }()
+
+			resultLun, resultErr := c.cloud.AttachDisk(ctx, asyncAttach, diskName, volumeID, nodeName, cachingMode, disk)
+			if resultErr != nil {
+				w.Logger().Errorf(err, "attach volume %q to instance %q failed", volumeID, nodeName)
+			}
+			w.Logger().V(2).Infof("attach operation successful: volume %q attached to node %q.", volumeID, nodeName)
+			attachResultCh <- azureutils.SetCloudAttachResult(map[string]string{azureconstants.LUN: strconv.Itoa(int(resultLun))}, resultErr)
```

Comment:
Do we really need to return the LUN as part of the `CloudAttachResult`? It should be the same as the one we already received from the `lunCh` and returned to the caller.

Reply:
Yes, that sounds right. I will address this along with an issue I spotted with blocking lun channel.